# Chapter 1: Python Basics for Financial Data Mining

This chapter introduces the working environment and the Python concepts needed
throughout the course. The examples use small finance-oriented calculations so
that syntax is learned together with interpretation.

## Learning map

1. Python environments and the Jupyter workflow
2. Variables, indexing, and case sensitivity
3. Lists, tuples, and mutability
4. Strings, comments, and mathematical modules
5. Functions and parameter passing
6. Future value and present value functions

**Code availability:** Yes. Every code cell is self-contained and can be run
independently.

## How to Study This Chapter

- Read the syntax rule and immediately trace the corresponding finance example.
- Re-type short functions instead of memorizing only their output.
- Pay special attention to indentation, indexing, mutability, and `return`.
- Before execution, predict the data type and printed value.

## Exam Checklist

- Can I distinguish a list, tuple, string, and numerical scalar?
- Can I define and call a function with positional and keyword arguments?
- Can I explain every symbol in the future-value and present-value formulas?


## 1.1 Python Tools Used in the Course

- **Spyder** is suitable for long scripts, debugging, and reusable programs.
- **Jupyter Notebook** is suitable for exploratory analysis, formulas, charts,
  explanations, and reproducible reports.
- **JupyterLab** provides notebooks, terminals, and file management in one
  browser-based workspace.

In Jupyter, press `Shift+Enter` to execute the current cell and move to the next
cell. Python is case-sensitive, uses indentation to define code blocks, and
starts sequence indexing at zero.


## 1.2 Variables, Case Sensitivity, and Zero-Based Indexing


A variable stores an object. `price` and `Price` are different names. A list
index identifies the position of an observation, beginning with index `0`.
This matters when selecting the first price, return, or cash flow.



### Structured Review

- **Object:** Identify the value and data type stored by each variable.
- **Access:** Remember that Python sequence indexing begins at zero.
- **Mutability:** Decide whether the object may be changed after creation.
- **Function contract:** Separate parameters, calculation steps, and the returned value.

### Exam Focus

- Preserve indentation inside functions and conditional blocks.
- Distinguish positional arguments from keyword arguments.
- Trace the value returned by a function for a small input.

### Common Mistakes

- Using the wrong capitalization in a variable name.
- Attempting to modify a tuple.
- Printing a result inside a function when the question requires `return`.


In [1]:
'''
本单元演示变量赋值、大小写敏感和从零开始的索引。
价格序列代表连续三个交易日的收盘价。
'''
price = 100
Price = 105
closing_prices = [100.0, 101.5, 99.8]

# Python 序列从 0 开始计数，索引 0 表示第一项
print("price =", price)
print("Price =", Price)
print("First closing price =", closing_prices[0])
print("Third closing price =", closing_prices[2])


price = 100
Price = 105
First closing price = 100.0
Third closing price = 99.8


**Result interpretation.** The two differently capitalized variables coexist, and index `0` returns the first market observation.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 1.3 Lists and Tuples


Lists are mutable: their elements can be updated after creation. Tuples are
immutable and are useful for values that should stay fixed, such as a contract
description or a pair of model parameters.



### Structured Review

- **Object:** Identify the value and data type stored by each variable.
- **Access:** Remember that Python sequence indexing begins at zero.
- **Mutability:** Decide whether the object may be changed after creation.
- **Function contract:** Separate parameters, calculation steps, and the returned value.

### Exam Focus

- Preserve indentation inside functions and conditional blocks.
- Distinguish positional arguments from keyword arguments.
- Trace the value returned by a function for a small input.

### Common Mistakes

- Using the wrong capitalization in a variable name.
- Attempting to modify a tuple.
- Printing a result inside a function when the question requires `return`.


In [2]:
'''
本单元比较列表与元组。
列表中的投资组合权重可以调整；元组中的合约信息保持不变。
'''
portfolio_weights = [0.60, 0.40]

# 列表可变：通过索引直接修改指定位置的元素
portfolio_weights[0] = 0.55
portfolio_weights[1] = 0.45

# 元组不可变，适合保存不应被修改的合约属性
bond_contract = ("US Treasury", 10)

print("Adjusted weights =", portfolio_weights)
print("Fixed contract description =", bond_contract)
print("Weight sum =", sum(portfolio_weights))


Adjusted weights = [0.55, 0.45]
Fixed contract description = ('US Treasury', 10)
Weight sum = 1.0


**Result interpretation.** The list accepts rebalancing and the weights still sum to one; the tuple represents fixed descriptive information.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 1.4 Strings and Mathematical Constants


Strings store labels such as ticker symbols. Methods like `.upper()` and
`.lower()` standardize text. The standard-library `math` module provides
constants and mathematical functions without requiring an external package.



### Structured Review

- **Object:** Identify the value and data type stored by each variable.
- **Access:** Remember that Python sequence indexing begins at zero.
- **Mutability:** Decide whether the object may be changed after creation.
- **Function contract:** Separate parameters, calculation steps, and the returned value.

### Exam Focus

- Preserve indentation inside functions and conditional blocks.
- Distinguish positional arguments from keyword arguments.
- Trace the value returned by a function for a small input.

### Common Mistakes

- Using the wrong capitalization in a variable name.
- Attempting to modify a tuple.
- Printing a result inside a function when the question requires `return`.


In [3]:
'''
本单元演示字符串标准化以及 math 模块中的常数和函数。
按需导入 math，避免依赖其他单元。
'''
import math

ticker = "ibm"

# math.exp(x) 计算 e 的 x 次方，常用于连续复利
continuous_growth = math.exp(0.05)

# upper() 将字符串统一转换为大写
print("Standardized ticker =", ticker.upper())
print("Euler's number =", round(math.e, 6))
print("Pi =", round(math.pi, 6))
print("One-year growth factor at 5% continuous growth =", round(continuous_growth, 6))


Standardized ticker = IBM
Euler's number = 2.718282
Pi = 3.141593
One-year growth factor at 5% continuous growth = 1.051271


**Result interpretation.** The ticker is standardized for consistent labeling, and `exp(0.05)` gives the continuous-compounding growth factor.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 1.5 Defining Reusable Functions


A function packages a calculation behind a clear name. Parameters are inputs,
and `return` sends the computed value back to the caller. Keyword arguments make
financial inputs easier to audit because their meanings are explicit.



### Structured Review

- **Object:** Identify the value and data type stored by each variable.
- **Access:** Remember that Python sequence indexing begins at zero.
- **Mutability:** Decide whether the object may be changed after creation.
- **Function contract:** Separate parameters, calculation steps, and the returned value.

### Exam Focus

- Preserve indentation inside functions and conditional blocks.
- Distinguish positional arguments from keyword arguments.
- Trace the value returned by a function for a small input.

### Common Mistakes

- Using the wrong capitalization in a variable name.
- Attempting to modify a tuple.
- Printing a result inside a function when the question requires `return`.


In [4]:
'''
本单元定义一个简单的收益放大函数。
函数、参数、缩进和 return 是后续金融模型复用的基础。
'''
def scale_return(period_return, multiplier):
    # return 将计算结果返回给函数调用者
    return period_return * multiplier

base_return = 0.025

print("Positional arguments =", scale_return(base_return, 2))
print("Keyword arguments =", scale_return(period_return=base_return, multiplier=3))


Positional arguments = 0.05
Keyword arguments = 0.07500000000000001


**Result interpretation.** The same function works with positional or keyword arguments; keyword input is usually clearer in longer financial formulas.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 1.6 Future Value


Future value compounds a present amount:

$$FV = PV(1+r)^n$$

The rate `r` and number of periods `n` must use the same frequency. For example,
an annual rate requires a number of years, while a monthly rate requires a
number of months.



### Structured Review

- **Inputs:** Identify the cash-flow amount, timing, period rate, and number of periods.
- **Timeline first:** Mark time 0 separately and determine whether payments occur at the beginning or end of each period.
- **Frequency check:** Convert the rate or period count whenever their frequencies do not match.
- **Decision link:** Separate a valuation result from a project decision rule.

### Exam Focus

- Be able to reconstruct the formula from the cash-flow timeline.
- Check the exponent, discounting direction, and cash-flow sign.
- Verify a library result with a short manual calculation whenever possible.

### Common Mistakes

- Mixing annual rates with monthly periods.
- Discounting the time-zero cash flow.
- Confusing an annuity due with an ordinary annuity.


In [5]:
'''
本单元封装复利终值函数。
pv 为现值，r 为每期利率，n 为期数；利率频率必须与期数一致。
'''
def future_value(pv, r, n):
    # 复利终值公式：现值乘以 (1 + 每期利率) 的 n 次方
    return pv * (1 + r) ** n

# 位置参数必须严格按照函数定义中的参数顺序传入
fv_positional = future_value(100, 0.10, 2)

# 关键字参数直接写明变量含义，顺序可以调整
fv_keyword = future_value(pv=100, r=0.0123, n=5)

print("FV of $100 at 10% for 2 years =", round(fv_positional, 2))
print("FV of $100 at 1.23% for 5 years =", round(fv_keyword, 2))


FV of $100 at 10% for 2 years = 121.0
FV of $100 at 1.23% for 5 years = 106.3


**Result interpretation.** Compounding produces $121.00 in the first example and approximately $106.30 in the lecture example.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 1.7 Present Value


Present value reverses compounding:

$$PV = \frac{FV}{(1+r)^n}$$

It answers how much a future cash flow is worth today at a required return or
discount rate.



### Structured Review

- **Inputs:** Identify the cash-flow amount, timing, period rate, and number of periods.
- **Timeline first:** Mark time 0 separately and determine whether payments occur at the beginning or end of each period.
- **Frequency check:** Convert the rate or period count whenever their frequencies do not match.
- **Decision link:** Separate a valuation result from a project decision rule.

### Exam Focus

- Be able to reconstruct the formula from the cash-flow timeline.
- Check the exponent, discounting direction, and cash-flow sign.
- Verify a library result with a short manual calculation whenever possible.

### Common Mistakes

- Mixing annual rates with monthly periods.
- Discounting the time-zero cash flow.
- Confusing an annuity due with an ordinary annuity.


In [6]:
'''
本单元封装现值函数，并用终值函数进行反向验证。
每个函数均在本单元内定义，因此可以独立运行。
'''
def future_value(pv, r, n):
    # 将现值向未来复利
    return pv * (1 + r) ** n

def present_value(fv, r, n):
    # 将未来值除以复利因子，折现回当前时点
    return fv / (1 + r) ** n

today = present_value(fv=110, r=0.10, n=1)

# 用终值函数反向验证现值计算是否正确
recovered_future = future_value(pv=today, r=0.10, n=1)

print("Present value =", round(today, 2))
print("Compounded back to the future =", round(recovered_future, 2))


Present value = 100.0
Compounded back to the future = 110.0


**Result interpretation.** Discounting $110 for one year at 10% gives $100, and compounding it again recovers the original future value.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 1.8 Lambda Functions from the Classroom Script


The instructor's source code also demonstrates `lambda`, a compact way to
define a short anonymous function. It is frequently used as an objective,
transformation, or row-level operation.



### Structured Review

- **Object:** Identify the value and data type stored by each variable.
- **Access:** Remember that Python sequence indexing begins at zero.
- **Mutability:** Decide whether the object may be changed after creation.
- **Function contract:** Separate parameters, calculation steps, and the returned value.

### Exam Focus

- Preserve indentation inside functions and conditional blocks.
- Distinguish positional arguments from keyword arguments.
- Trace the value returned by a function for a small input.

### Common Mistakes

- Using the wrong capitalization in a variable name.
- Attempting to modify a tuple.
- Printing a result inside a function when the question requires `return`.


In [7]:
'''
本单元补充老师源代码中的 lambda 函数。
lambda 后写参数，冒号后写返回表达式，适合简短的一行函数。
'''
# 单变量 lambda：输入收益率并转换为增长因子
growth_factor = lambda return_rate: 1 + return_rate

# 多变量 lambda：x 是包含两个元素的向量
distance_objective = lambda x: (x[0] - 10) ** 2 + (x[1] - 25) ** 2

print("Growth factor =", growth_factor(0.08))
print("Objective at (11, 27) =", distance_objective((11, 27)))


Growth factor = 1.08
Objective at (11, 27) = 5


**Result interpretation.** The first lambda returns 1.08. The second evaluates squared distance from the target point `(10, 25)`.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## Chapter Review

- Python names are case-sensitive and indentation defines program structure.
- Sequences start at index zero.
- Lists are mutable; tuples are immutable.
- Modules provide reusable functions and constants.
- Functions turn formulas into auditable, reusable tools.
- Financial rates and periods must always use matching frequencies.
